testscripts/9_regular_user_can_burn_his_tokens.mvir

```diff
@@ -0,0 +1,378 @@
+        let is_minter: bool;
+        is_minter = *(&move(capability).minter);
+        return move(is_minter);
+    }
+
+    // Return true the capability is not blacklisted.
+    public is_not_blacklisted(capability: &R#Self.T): bool {
+        let is_blacklisted: bool;
+        is_blacklisted = *(&move(capability).blacklisted);
+        return !move(is_blacklisted);
+    }
+
+    // Reverts if capability does not allow minting
+    public require_minter(capability: &R#Self.T) {
+        let is_minter: bool;
+        is_minter = Self.is_minter(move(capability));
+        assert(move(is_minter), 1);
+        return;
+    }
+
+    // Reverts if capability is blacklisted
+    public require_not_blacklisted(capability: &R#Self.T) {
+        let is_not_blacklisted: bool;
+        is_not_blacklisted = Self.is_not_blacklisted(move(capability));
+        assert(move(is_not_blacklisted), 2);
+        return;
+    }
+}
+
+//! new-transaction
+//! sender: alice
+// Alice publishes the module EToken
+
+module EToken {
+
+    // This module is responsible for an actual eToken.
+    // For it to be useful a capability has to be published by using the Capability module above.
+
+    // -----------------------------------------------------------------
+
+    import {{alice}}.Capability;
+
+    // Declare the eToken resource, storing an account's total balance.
+    resource T {
+        value: u64,
+    }
+
+    // Publishes an initial zero eToken to the sender.
+    // Should be called once before using this module.
+    public publish() {
+        move_to_sender<T>(T{ value: 0 });
+        return;
+    }
+
+    // Mint new eTokens.
+    // Reverts if capability does not allow it.
+    public mint(value: u64, capability: &R#Capability.T): R#Self.T {
+        Capability.require_minter(move(capability));
+        return T{value: move(value)};
+    }
+
+    // Burns owned tokens
+    // TODO: decide whether capability should be passed to revert burns if user is blacklisted.
+    public burn(amount: u64) {
+        let sender: address;
+        let sender_token_ref: &mut R#Self.T;
+        let value: u64;let payee_token_ref: &mut R#Self.T;
+
+        // Getting sender-accounts balance
+        sender = get_txn_sender();
+        sender_token_ref = borrow_global<T>(move(sender));
+        value = *(&copy(sender_token_ref).value);
+
+        // Make sure that sender has enough tokens, otherwise revert
+        assert(copy(value) >= copy(amount), 1);
+
+        // Reduce the senders tokens
+        *(&mut move(sender_token_ref).value) = move(value) - copy(amount);
+
+        return;
+    }
+
+    // Burns owned tokens
+    // Fails if the value is zero
+    public burn_by_resource(to_burn: R#Self.T) {
+        let burned_value: u64;
+
+        // Unpack and destroy to_burn tokens
+        T { value: burned_value } = move(to_burn);
+
+        // Making sure more than 0 is burned
+        assert(move(burned_value) != 0, 3);
+        return;
+    }
```

Comment:
Burning by resource here not sure if the != 0 assert is necessary. It is less code than the other burn method but I am not sure what is better in regards to security and performance. 
I have added a basic test for this function on line 363

Reply:
The assertion shouldn't be necessary. I like this version better as it is more concise. It also seems more secure as the type system now guarantees that you can only burn what you own.